ui/src/app/base/components/MainTable/MainTable.js

```diff
@@ -0,0 +1,209 @@
+import PropTypes from "prop-types";
+import React, { useEffect, useState } from "react";
+
+import Table from "../Table";
+import TableRow from "../TableRow";
+import TableHeader from "../TableHeader";
+import TableCell from "../TableCell";
+
+const _updateSort = (
+  _setSortKey,
+  _setSortDirection,
+  sortKey,
+  sortDirection
+) => {
+  let newDirection = null;
+  if (sortDirection === "none") {
+    newDirection = "ascending";
+  } else if (sortDirection === "ascending") {
+    newDirection = "descending";
+  } else {
+    sortKey = null;
+  }
+  _setSortKey(sortKey);
+  _setSortDirection(newDirection);
+};
+
+const _generateHeaders = (
+  currentSortKey,
+  currentSortDirection,
+  expanding,
+  headers,
+  sortable,
+  _setSortKey,
+  _setSortDirection
+) => {
+  const headerItems = headers.map(({ content, sortKey, ...props }, index) => {
+    let sortDirection;
+    if (sortable && sortKey) {
+      if (currentSortKey === sortKey) {
+        sortDirection = currentSortDirection;
+      } else {
+        sortDirection = "none";
+      }
+    }
+    return (
+      <TableHeader
+        key={index}
+        sort={sortDirection}
+        onClick={
+          sortable &&
+          _updateSort.bind(
+            this,
+            _setSortKey,
+            _setSortDirection,
+            sortKey,
+            sortDirection
+          )
+        }
+        {...props}
+      >
+        {content}
+      </TableHeader>
+    );
+  });
+  // When there is expanding content then provide an extra hidden header to
+  // account for the extra cell in the body rows.
+  return (
+    <thead>
+      <TableRow>
+        {headerItems}
+        {expanding && <TableHeader className="u-hide"></TableHeader>}
+      </TableRow>
+    </thead>
+  );
+};
```

Comment:
Tiny nit, but nice to have a line between functions.

Reply:
Oops, thanks.